src/EventListener/SessionListener.php

```diff
@@ -73,9 +76,18 @@ public function onKernelResponse(FilterResponseEvent $event)
         // Call default behaviour if it's an irrelevant request for the user context
         if (0 === count(array_intersect($varyHeaders, $relevantHeaders))) {
             $this->inner->onKernelResponse($event);
+
+            return;
+        }
+
+        if (version_compare(Kernel::VERSION, '4.1', '<')) {
+            // For Symfony 3.4 and 4.0, do not let the inner listener destroy caching. The session will be saved in a separate listener.
+            return;
         }
 
-        // noop, see class description
+        // For Symfony 4.1+, we need to call the inner listener to save the session. Set the header to disable overwriting caching headers.
+        $event->getResponse()->headers->set(AbstractSessionListener::NO_AUTO_CACHE_CONTROL_HEADER, 1);
```

Comment:
I don't think this should be set here. I think we should not decorate the listener at all in Symfony 4.1+ but instead set that header where it belongs: The `UserContextListener`. Imho it can always be added, even for lower Symfony versions because the header doesn't harm at all if it is being sent to external but if you really want to omit that you can add the version_compare there.

So basically imho we should only register/decorate this listener when Symfony 3.4 or Symfony 4.0.
And then I think - for the sake of your time and stress level - you should really think about dropping < 4.1 support in the future 😄

Reply:
> So basically imho we should only register/decorate this listener when Symfony 3.4 or Symfony 4.0.

Yep, this goes inline with my suggestion to keep the version check at one place. It can be further extended to only decorate on Sf 3.4 and 4.0.

> And then I think - for the sake of your time and stress level - you should really think about dropping < 4.1 support in the future

I hope this won't happen before 3.0, though :)